### ISSUE TITLE
A2 - Potential /buy and /update issues

### ISSUE THREAD

**Issue by ellamgriffin** [CISC-CMPE-327/Information#60]
Hi there,

**Observation 1) Potential issue with logic in /buy route (identifying a ticket instance)**

- The form submitting a POST request to /buy consists of the fields 'name' and 'quantity' and performs the following two implicit actions:
1. Modifies the corresponding unique ticket instance stored in the database by updating the quantity field to subtract the quantity specified in the quantity input field of the form (i.e. reflecting a reduced amount of available tickets for the ticket specified by the name input field value)
2. Deducts the ticket price from the use balance

This would imply that a ticket instance can be UNIQUELY identified by the **name** attribute. For example, in this update operation, UPDATE db_name.Ticket set quantity = original quantity - request.form.get('ticket-quantity') where name = request.form.get('ticket-name').

I would argue that name is **not** a unique identifier of ticket. For example, owner x could be selling 2 'Leafs vs. Capitals Hockey Game' tickets for $50 and owner y could be selling 1 'Leafs vs. Capitals Hockey Game' ticket for $30.

Since the form posted to route /buy only has fields name and quantity, performing the query above could modify several ticket instances unless the constraint limit 1 is appended to the end of the query.

This could be resolved by having an additional field "owner email" added to the form to uniquely identify a ticket via a ticket name and owner.

**Observation 2) Potential issue with logic in /update route**

- The form submitting a POST request to /update consists of the fields 'name', 'quantity', 'price', 'expiration date' and performs the following action:

1. Modifies the corresponding unique ticket instance stored in the database by setting the name, quantity, price, and expiration date attributes to the corresponding values in the submitted form.

A constraint on this requirement is that "the ticket of the given name must exist". The above action and constraint imply:
1. That the name can uniquely identify a ticket
2. The name field in the form is immutable (can not be changed)

As for 1, I would argue that a ticket can not be uniquely identified by name alone. The only way of uniquely identifying a ticket would be via a unique id (primary key of Ticket table) or a composite key on all other attributes/columns (including owner email). Perhaps each list item can be a link (uniquely identified by the ticket_id) and on select, the form associated with /update can be populated with these fields and a new field "Ticket Id: " can be displayed as readonly/disabled and passed as a parameter to the post request to be specified as a constraint in the update query?

As for 2, I think it is realistic to be able to change the name of a ticket. If this ability is allowed (as it does not explicitly say that a ticket name can not be changed), it would directly violate the constraint and yield an error on submit.

If a ticket name can not be changed, then we cycle back to the existing issue of name not uniquely identifying a ticket.

I believe sub-requirements of R5 (/buy) and R6 (/update) are contradictory or at least, not logically sound.

Please reply soon as this directly affects implementation requirements for A2 due 11/07.

Thank you!

**Comment by leordsong** [comment]
Is this for web or CLI?

**Comment by ellamgriffin** [comment]
Sorry...web

**Comment by leordsong** [comment]
Please read assignment requirements carefully. R4, R5 and R6 are not required for A2.

**Comment by leordsong** [comment]
Yes, you need an identifier for a ticket. It can be string or integer.

**Comment by ellamgriffin** [comment]
Should we not implement the backend DB logic for /buy /sell and /update then?

**Comment by leordsong** [comment]
@ellamgriffin Yes, that's correct (You don't need to implement backend for /buy /sell and /update).
Sorry about the confusion for **R3**. I believe you can skip the ticket part for **R3**. Let me confirm with @steven-hh-ding .
